homeassistant/components/zwave_js/climate.py

```diff
@@ -0,0 +1,303 @@
+    ThermostatMode.COOLING_ECON: HVAC_MODE_COOL,
+    ThermostatMode.AWAY: HVAC_MODE_HEAT_COOL,
+    ThermostatMode.FULL_POWER: HVAC_MODE_HEAT,
+}
+
+HVAC_CURRENT_MAP: Dict[int, str] = {
+    ThermostatOperatingState.IDLE: CURRENT_HVAC_IDLE,
+    ThermostatOperatingState.PENDING_HEAT: CURRENT_HVAC_IDLE,
+    ThermostatOperatingState.HEATING: CURRENT_HVAC_HEAT,
+    ThermostatOperatingState.PENDING_COOL: CURRENT_HVAC_IDLE,
+    ThermostatOperatingState.COOLING: CURRENT_HVAC_COOL,
+    ThermostatOperatingState.FAN_ONLY: CURRENT_HVAC_FAN,
+    ThermostatOperatingState.VENT_ECONOMIZER: CURRENT_HVAC_FAN,
+    ThermostatOperatingState.AUX_HEATING: CURRENT_HVAC_HEAT,
+    ThermostatOperatingState.SECOND_STAGE_HEATING: CURRENT_HVAC_HEAT,
+    ThermostatOperatingState.SECOND_STAGE_COOLING: CURRENT_HVAC_COOL,
+    ThermostatOperatingState.SECOND_STAGE_AUX_HEAT: CURRENT_HVAC_HEAT,
+    ThermostatOperatingState.THIRD_STAGE_AUX_HEAT: CURRENT_HVAC_HEAT,
+}
+
+
+async def async_setup_entry(
+    hass: HomeAssistant, config_entry: ConfigEntry, async_add_entities: Callable
+) -> None:
+    """Set up Z-Wave climate from config entry."""
+    client: ZwaveClient = hass.data[DOMAIN][config_entry.entry_id][DATA_CLIENT]
+
+    @callback
+    def async_add_climate(info: ZwaveDiscoveryInfo) -> None:
+        """Add Z-Wave Climate."""
+        entities: List[ZWaveBaseEntity] = []
+        entities.append(ZWaveClimate(client, info))
+
+        async_add_entities(entities)
+
+    hass.data[DOMAIN][config_entry.entry_id][DATA_UNSUBSCRIBE].append(
+        async_dispatcher_connect(
+            hass, f"{DOMAIN}_add_{CLIMATE_DOMAIN}", async_add_climate
+        )
+    )
+
+
+class ZWaveClimate(ZWaveBaseEntity, ClimateEntity):
+    """Representation of a Z-Wave climate."""
+
+    def __init__(self, client: ZwaveClient, info: ZwaveDiscoveryInfo) -> None:
+        """Initialize lock."""
+        super().__init__(client, info)
+        self._hvac_modes: Dict[str, Optional[int]] = {}
+        self._hvac_presets: Dict[str, Optional[int]] = {}
+        for enum in ThermostatSetpointType:
+            self.get_zwave_value(
+                THERMOSTAT_SETPOINT_PROPERTY,
+                command_class=CommandClass.THERMOSTAT_SETPOINT,
+                value_property_key_name=enum.value,
+                add_to_watched_value_ids=True,
+            )
+        self.get_zwave_value(
+            THERMOSTAT_OPERATING_STATE_PROPERTY,
+            command_class=CommandClass.THERMOSTAT_OPERATING_STATE,
+            add_to_watched_value_ids=True,
+        )
+        self.get_zwave_value(
+            THERMOSTAT_CURRENT_TEMP_PROPERTY,
+            command_class=CommandClass.SENSOR_MULTILEVEL,
+            add_to_watched_value_ids=True,
+        )
+
+        self.on_value_update()
+
+    @callback
+    def on_value_update(self) -> None:
+        """Call when a watched value is added or updated."""
+        self._current_mode_setpoint_values = self._get_current_mode_setpoint_values()
```

Comment:
I think it should be enough to get the setpoint values during entity set up. We don't need to do it every value update.

The entity won't be added until the node is ready. The node is ready when it has completed the interview at least once.

https://zwave-js.github.io/node-zwave-js/#/api/node?id=quotreadyquot

Reply:
we need to do this specifically anytime the `hvac_mode` changes. Right now this is being done too often but moving it into initialization will not be enough

Reply:
The available modes are static after the interview is done, right? So we should be able to store the set point values for each mode in a dictionary when setting up the entity. Then afterwards we can quickly look up the setpoint based on current mode.

Reply:
sorry I misspoke, the reason we need to check for the setpoint value on update is because the setpoint value can change. The function `_get_current_mode_setpoint_values` is using the `THERMOSTAT_MODE_SETPOINT_MAP` constant to look up which setpoints to check, so we are already quickly looking up the setpoints based on current mode, but we still need to check the setpoint values when there's an update.

Reply:
As long as we track all the setpoint values we will get a state update triggered when any of those values change. So it should be enough to look up the proposed dict for the correct list of setpoint values, depending on current mode, in the entity state properties and access the correct value.